src/main/java/team9/ddang/member/service/FriendServiceImpl.java

```diff
@@ -0,0 +1,119 @@
+    private final MemberDogRepository memberDogRepository;
+    private final WalkRepository walkRepository;
+    private final WalkWithMemberRepository walkWithMemberRepository;
+
+    @Override
+    @Transactional
+    public MemberResponse addFriend(Member member, AddFriendRequest addFriendRequest) {
+        Member otherMember = getMemberFromMemberIdOrElseThrow(addFriendRequest.memberId());
+
+        if(friendRequestRepository.existsFriendRequestByReceiver(member)){
+            saveFriend(member, otherMember);
+        }
+        else {
+            createFriendRequest(member, otherMember);
+        }
+
+        return MemberResponse.from(otherMember);
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public List<FriendListResponse> getFriendList(Member member) {
+        List<Member> friends = friendRepository.findAllFriendsBySender(member);
+
+        return friends.stream().map(FriendListResponse::from).toList();
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public FriendResponse getFriend(Member member, Long memberId) {
+        Member otherMember = getMemberFromMemberIdOrElseThrow(memberId);
+
+        if(!friendRepository.existsBySenderAndReceiver(member, otherMember)){
+            throw new IllegalArgumentException("친구가 아닌 사람의 프로필은 볼 수 없습니다.");
+        }
+
+        Dog dog = memberDogRepository.findMemberDogByMemberId(memberId)
+                .orElseThrow(() -> new IllegalArgumentException("강아지가 존재하지 않습니다.")).getDog();
+
+        int totalDistanceInMeters = walkRepository.findTotalDistanceByMemberId(memberId);
+        int countWalks = walkRepository.countWalksByMemberId(memberId);
+        double totalDistanceInKilometers = totalDistanceInMeters / 1000.0;
+        int countWalksWithMember = walkWithMemberRepository.countBySenderMemberId(memberId);
+
+        return FriendResponse.of(otherMember, dog, totalDistanceInKilometers, countWalks ,countWalksWithMember);
+    }
+
+    @Override
+    @Transactional
+    public void deleteFriend(Member member, Long memberId) {
+        Member otherMember = getMemberFromMemberIdOrElseThrow(memberId);
+
+        if(!friendRepository.existsBySenderAndReceiver(member, otherMember)){
+            throw new IllegalArgumentException("친구가 아닌 사람을 삭제할 수 없습니다.");
+        }
+
+        friendRepository.deleteBySenderAndReceiver(member, otherMember);
+        friendRepository.deleteBySenderAndReceiver(otherMember, member);
+    }
+
+    private void saveFriend(Member member, Member otherMember){
+        Friend friend = Friend.builder()
+                .sender(member)
+                .receiver(otherMember)
+                .build();
+
+        Friend friend2 = Friend.builder()
+                .sender(otherMember)
+                .receiver(member)
+                .build();
+
+        friendRepository.save(friend);
+        friendRepository.save(friend2);
+        friendRequestRepository.deleteByReceiver(member);
+    }
```

Comment:
중복 로직을 생각하면 다음과 같이 작성할 수 있을 것 같아요
```suggestion
    private void saveFriend(Member member, Member otherMember){
        List<Friend> friends = List.of(
            Friend.builder().sender(member).receiver(otherMember).build(),
            Friend.builder().sender(otherMember).receiver(member).build()
        );
        friendRepository.saveAll(friends);
        friendRequestRepository.deleteByReceiver(member);
    }
```

Reply:
아하 넵 참고해서 작성하겠습니다